Launchpad: Allows users to go back to Setup from Launchpad

edanzer opened the pull request:
### Proposed Changes

Related to #https://github.com/Automattic/wp-calypso/issues/67429

We want to users to return to the site Setup screen from Launchpad, and to return directly to Launchpad after making changes (rather than going through intermediate steps). Changes made: 
- To simplify code, I refactored to pass the 'navigation' object rather than passing its 'submit', 'goNext', and 'goToStep' properties/methods separately. 
- Added 'keepActive' property to the newsletterSetup and linkInBioSetup steps, so they will always remain clickable. 
- Added actionDispatch properties/methods to newsletterSetup and newsletterSetup that will direct users to the relevant setup screens. 
- Updated the Setup screens to look check if backToPage === launchpad. If so, it means they came from launchpad, and we will direct users directly back to launchpad. If not, it means they are on the Setup screen for the first time and we directly them to the next step as before. 

### Testing Instructions

Check out this branch and run yarn and yarn start if needed. You'll then want to test that this works with both Newsletter and Link in Bio flows. 

Test Newsletter flow: 
1) Go to http://calypso.localhost:3000/setup/launchpad?flow=newsletter&siteSlug=YOUR_SITE_SLUG
2) Confirm that Set up Newsletter is now clickable. It should still show as complete but should highlight blue on hover, and be clickable. 
3) Click that task and confirm it takes you back to the Newsletter setup screen (with site title, description, and icon). 
4) From the Setup screen, click the Continue button, and confirm you are taken directly back to the Launchpad screen. You should not go to the domains screen or any other intermediate steps before Launchpad. 

Test Link in Bio flow: 
1) Go to http://calypso.localhost:3000/setup/launchpad?flow=link-in-bio&siteSlug=YOUR_SITE_SLUG
2) Confirm that Set up Link in Bio is now clickable. It should still show as complete but should highlight blue on hover, and be clickable. 
3) Click that task and confirm it takes you back to the Link in Bio setup screen (with site title, description, and icon). 
4) From the Setup screen, click the Continue button, and confirm you are taken directly back to the Launchpad screen. You should not go to the domains screen or any other intermediate steps before Launchpad.

edanzer committed:
Enable link to Setup step from Launchpad

edanzer committed:
Allow direct return to Launchpad from Setup step

edanzer committed:
Apply new keepActive property to Setup pages

edanzer committed:
Add return-to-launchpa logic to Link-in-bio setup

Addison-Stavlo reviewed (commented):
Awesome, thanks for starting on this! I think navigation wise this works well and makes a lot of sense.

However, I think we might require more changes for the setup page here to apply any changes made.  Right now if I go back to the setup page and change information (add a logo, update description, etc.), and hit "continue", none of these changes seem to be reflected on my site.

I think this might be because this page is set up to take this info before site creation, and then use that when creating the site. Now, the site is already created and we aren't actually doing anything with what is added in the form this way?

Also, more minor but nice to fix... when I revisit the setup page from an existing site for the first time the title/description (and maybe logo? mine was never set originally) are all blank. It would be good if these pre-populated with the actual title, description, and logo of the site.

sharpirate commented on client/landing/stepper/declarative-flow/internals/steps-repository/link-in-bio-setup/index.tsx:
```diff
@@ -92,7 +94,8 @@ const LinkInBioSetup: Step = function LinkInBioSetup( { navigation } ) {
 		if ( siteTitle.trim().length ) {
-			submit?.( { siteTitle, tagline } );
+			const backToStep = urlQueryParams.get( 'backToStep' );
+			submit?.( { siteTitle, tagline, backToStep } );
```
do we need to check if backToStep is undefined here? does it work if it's undefined but we still pass it to the function?

edanzer replied:
@Sharpirate - Good eyes on this - I had checked the same thing. If the query param is not set, this returns null, not undefined. So we'd be passing a null value, and that works OK downstream for the needed conditional check.

sharpirate commented on client/landing/stepper/declarative-flow/internals/steps-repository/launchpad/task-helper.ts:
```diff
@@ -32,6 +31,11 @@ export function getEnhancedTasks(
 					taskData = {
 						title: translate( 'Set up Newsletter' ),
+						keepActive: true,
+						actionDispatch: () =>
```
At this point we are using actionDispatch for almost every case, only 'first-post-published' and 'links-added' are using actionUrl to perform the redirect, currently they don't exist on the StepPath type. Do you think we should add them there and use only actionDispatch with the goToStep method or we should keep using actionUrl

edanzer replied:
@Sharpirate - First post published is outside the stepper path as you noted, so we wouldn't use one of the stepper methods. That said, we could plausibly use actionDispatch, and just refactor downstream to use that method rather than actionUrl. There may be some advantage to just using that for all tasks / links. But I'd probably consider that a separate PR to remove actionUrl here and from the ChecklistItem component.

For the Links Added, I think that may be updated to use actionDispatch when we work through [this issue ](https://github.com/Automattic/wp-calypso/issues/67430https://github.com/Automattic/wp-calypso/issues/67430), which I think is one you'll be taking...

agrullon95 reviewed (approved):
Tested and I was able to navigate to link-in-bio/newsletter setup screens then back to Launchpad. The same values seemed to be saved and pre-populated on the set up screen even after I created a new site. We should figure out a way to maybe use redux to retrieve/set these setup screen values.

edanzer committed:
Ensure Setup screens prepopulate site title & tagline

edanzer committed:
Ensure site title/tagline are updated

simison commented:
> However, I think we might require more changes for the setup page here to apply any changes made. Right now if I go back to the setup page and change information (add a logo, update description, etc.), and hit "continue", none of these changes seem to be reflected on my site.
>
> I think this might be because this page is set up to take this info before site creation, and then use that when creating the site. Now, the site is already created and we aren't actually doing anything with what is added in the form this way?

It would be worth evolving `/setup` towards where site's existence in a way doesn't matter to any individual steps. If the site exists, it's a kind of "live editor" which modifies the site as you would in the editor. If the site doesn't exist, the info is collected to a state and used for creating the new site.

An alternative would be creating the site always earlier, but I'd rather see us move conceptually to other direction where everything can be "headless" or "account-less" some day until they need to hit save/publish in the editor.

cc @Automattic/vertex in case any framework level thoughts.

alshakero reviewed (changes requested):
This is a tiny and clean change, but I think it's not down a good path tbh.

The main Stepper's principle (I should really write these down) is that steps are just fancy form components that either submit stuff, or store them in a store. They shouldn't interact with each other at all (like form inputs), and should only do so when absolutely necessary. I even regret introducing `goBack` and `goNext` in `NavigationControls` and if I were to start fresh I'd only add `submit`. 

This PR takes Stepper down the /start path, where the steps are spaghetti and impossible to track and tell which is connected to which. Especially using query parameters, which are untyped, untraceable, and hard to document.

My solution to customization steps from Launchpad would be to spin new flows! Create a new micro flow similar to [this](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/link-in-bio.ts#L0-L1) under a special URL like `post-link-in-bio` and redirect to it from Launchpad. 

This will add even more incentive to make the site setup step generic to work in both flows. The step would then submit the answers, and the flow decides what to do with them. 

I'm aware this is the second time when I come and derail a Cylon PR, I promise I love your work and don't mean to, it's just a coincidence 🙏🏼 .

simison commented:
>I even regret introducing goBack and goNext in NavigationControls and if I were to start fresh I'd only add submit.

I think it's still early enough and manageable to do some refactoring like that. It's normal to do learnings from the actual use only after some time.

edanzer commented:
> My solution to customization steps from Launchpad would be to spin new flows!

Hey @alshakero - No worries and thanks for the input! To clarify, would your preference be that there's no navigation backwards within the stepper flow? 

I ask because my understanding has been that navigation between steps is fairly central to launchpad as currently conceived. We already added navigation back to prior steps [here](https://github.com/Automattic/wp-calypso/pull/67405) and [here](https://github.com/Automattic/wp-calypso/pull/67755). And if we added a forward step to a different setup-like screen, we'd still need to allow them to come backward to Launchpad again (and likely back and forth more than once).

edanzer commented:
@alshakero Yes, this makes sense and I understand the distinction now. Happy to adjust direction on this. Given the nature of the changes, I may close this PR, open new ones with the new work. and let you know when those are ready for review.

edanzer commented:
Closing this in favor of https://github.com/Automattic/wp-calypso/pull/67835
